src/Init/Data/BitVec/Lemmas.lean

```diff
@@ -2030,6 +2030,39 @@ theorem eq_msb_cons_setWidth (x : BitVec (w+1)) : x = (cons x.msb (x.setWidth w)
   ext i
   simp [cons]
 
+
+theorem cons_append (x : BitVec w₁) (y : BitVec w₂) (a : Bool) :
+    (cons a x) ++ y = (cons a (x ++ y)).cast (by omega) := by
+  ext i h
+  simp only [cons, getLsbD_append, getLsbD_cast, getLsbD_ofBool, cast_cast]
+  by_cases h₀ : i < w₁ + w₂
+  · simp [h₀]
+    by_cases h₁ : i < w₂
+    · simp [h₁]
+    · simp [h₁, show i - w₂ - w₁ = 0 by omega]
+      omega
+  · simp [show ¬i < w₂ by omega, show i - w₂ - w₁ = 0 by omega, h₀, show i - (w₁ + w₂) = 0 by omega]
+    omega
```

Comment:
I'm sorry to keep asking for more and more changes here, but I think there's opportunity to fill in obvious missing lemmas and get easier proofs.

Could this be by:
```
theorem cons_append (x : BitVec w₁) (y : BitVec w₂) (a : Bool) :
    (cons a x) ++ y = (cons a (x ++ y)).cast (by omega) := by
  apply eq_of_toNat_eq
  simp only [toNat_append, toNat_cons, toNat_cast]
  rw [Nat.shiftLeft_add, Nat.shiftLeft_or_distrib]
```
The lemma `Nat.shiftLeft_or_distrib` is missing and would need to be written.

Reply:
The idea here is that switching to Nat as soon as possible means you don't need to handle all in index inequalities. It's great that `omega` can solve them, but hopefully we can just avoid them to begin with.

Reply:
Thanks for the advice @kim-em  :) I'm working on the proof of `Nat.shiftLeft_or_distrib` but it seems to me like the best way to prove it is to add `shiftLeft_bitwise_distrib`, `bitwise_mul_two_pow`, and potentially a few more theorems (basically do what the proof of `shiftRight_and_distrib` does). If this sounds sensible I'm happy to add the necessary theorems, but maybe it's better to do it in another PR, since this is quite big already?

Reply:
Currently working [here](https://github.com/opencompl/lean4/tree/nat-shiftleft-distrib-s) on this to avoid having a massive PR!

Reply:
now in #6630 :)